File: Chapter - 2 -  Linked List/2. Return Kth to last/ReturnKthToLast.cpp
#include <iostream>
using namespace std;

struct Node{
	int data;
	Node* next;
	Node(int x){
		data = x;
		next = NULL;
	}
};

Node* pushToLast(Node* head, int data){
	Node* node = new Node(data);
	Node* temp = head;
	if(head == NULL){
		head = node;
	}else{
		while(temp && temp->next) temp = temp->next;
		temp->next = node;
	}
	return head;
}

Node* pushToFront(Node* head, int data){
	Node* node = new Node(data);
	Node* temp = head;
	if(temp == NULL) head = node;
	else{
		node->next = head;
		head = node;
	}
	return head;
}

int getLength(Node* head){
	Node* temp = head;
	int count = 0;
	while(temp){
		count++;
		temp = temp->next;
	}
	return count;
}

void printList(Node* head){
	Node* temp = head;
	while(temp){
		cout << temp->data << " ";
		temp = temp->next;
	}
}

void kthToLast(Node *head){
	int k;
	cin >> k;
	int n = getLength(head);
	if(k > n){
		cout << "k is greater than n !!!!";
	}else{
		Node* slow = head;
		Node* fast = head;
		while(k--) fast = fast->next;
		while(fast){
			slow = slow->next;
			fast = fast->next;
		}
		cout << "kth from last is : " << slow->data;
	}
}

int main() {
	//Number of nodes in the linked list
	int n;
	cin >> n;
	Node* head = NULL;
	for(int i = 0 ; i < n ; i++){
		int x;
		cin >> x;
		head = pushToLast(head, x);
	}
	cout << "Old list was : ";
	printList(head);
	cout << endl;
	
	kthToLast(head);
	
	return 0;
}

File: Chapter - 2 -  Linked List/2. Return Kth to last/ReturnKthToLast.java
class ReturnKthToLast {
    static ListNode head = null;

    public static ListNode kthLast(ListNode head, int k) {
        ListNode slow = head;
        ListNode fast = head;

        while (k > 0 && fast != null) {
            fast = fast.next;
            k--;
        }

        while (fast != null) {
            fast = fast.next;
            slow = slow.next;
        }

        return slow;
    }

    public static void print(ListNode head) {
        ListNode current = head;
        while (current != null) {
            System.out.print(current.val + " ");
            current = current.next;
        }
        System.out.println();
    }

    public static void add(int data) {
        if (head == null) {
            head = new ListNode(data);
            return;
        }
        ListNode curr = head;
        while (curr.next != null) {
            curr = curr.next;
        }
        curr.next = new ListNode(data);
    }

    public static void main(String[] args) {
        add(3);
        add(5);
        add(31);
        add(67);
        add(8);
        add(90);
        add(43);

        print(head);
        System.out.println("Kth Last: " + kthLast(head, 5).val);
    }

    static class ListNode {
        int val;
        ListNode next;

        ListNode(int x) {
            val = x;
        }
    }
}